gotfparse/pkg/converter/converter.go

```diff
@@ -36,118 +30,214 @@ type terraformMeta struct {
-					refIdJ, _ := json.Marshal(rb.ID())
-					refIdP, _ := gabs.ParseJSON(refIdJ)
-					obj[a.Name()] = refIdP
-				}
-			}
+			obj[a.Name()] = toRawValue(a)
+		}
+	}
 
-			idJ, _ := json.Marshal(b.ID())
-			id, _ := gabs.ParseJSON(idJ)
+	id := b.ID()
+	if id != "" {
+		obj["id"] = id
+	}
 
-			meta := generateTFMeta(b)
-			if meta != nil {
-				obj["__tfmeta"] = meta
-			}
+	obj["__tfmeta"] = generateTFMeta(b)
 
-			if id != nil {
-				obj["id"] = id
-			}
+	return obj
+}
+
+func (t *terraformConverter) visitBlock(b *terraform.Block, parentPath string, jsonOut *gabs.Container) {
+	arrayKey := t.getPath(b, parentPath)
 
-			if len(obj) > 0 {
-				blockJSON.SetP(obj, arrayKey)
+	if !t.shouldWalk(b) {
+		return
+	}
+
+	if t.isResource(b) {
+		json := t.buildBlock(b)
+		json["__tfmeta"].(map[string]interface{})["path"] = arrayKey
+		jsonOut.ArrayAppendP(json, b.TypeLabel())
+	}
+
+	process := handleDynamicBlocks(func(block *terraform.Block) {
+		t.visitBlock(block, arrayKey, jsonOut)
+	})
+	for _, child := range b.AllBlocks() {
+		process(child)
+	}
+}
+
+func toRawValue(a *terraform.Attribute) interface{} {
+	val := a.Value()
+
+	if raw, ok := fromValueToRawValue(val); ok {
+		return raw
+	}
+
+	return a.GetRawValue()
+}
+
+func fromValueToRawValue(val cty.Value) (interface{}, bool) {
+	var (
+		ok bool
+
+		vType = val.Type()
+	)
+
+	if val.IsNull() {
+		return nil, true
+	}
+
+	if !val.IsKnown() {
+		return nil, true
+	}
+
+	if vType.IsObjectType() || vType.IsMapType() {
+		valueMap := val.AsValueMap()
+		interfaceMap := make(map[string]interface{})
+		for key, val := range valueMap {
+			if interfaceMap[key], ok = fromValueToRawValue(val); !ok {
+				return nil, false
+			}
+		}
+		return interfaceMap, true
+	}
+	if vType.IsListType() || vType.IsTupleType() {
+		valueSlice := val.AsValueSlice()
+		interfaceSlice := make([]interface{}, len(valueSlice))
+		for idx, item := range valueSlice {
+			if interfaceSlice[idx], ok = fromValueToRawValue(item); !ok {
+				return nil, false
 			}
 		}
+		return interfaceSlice, true
+	}
+
+	if vType == cty.String {
+		return val.AsString(), true
+	}
 
-		if jsonOut.ExistsP(arrayKey) {
-			data := blockJSON.Path(arrayKey)
-			jsonOut.ArrayAppendP(data, arrayKey)
+	if vType == cty.Number {
+		num := val.AsBigFloat()
+		if num.IsInt() {
+			i, _ := num.Int64()
+			return i, true
 		} else {
-			jsonOut.MergeFn(blockJSON, collisionFn(arrayKey))
+			f, _ := num.Float64()
+			return f, true
 		}
+	}
 
-		for _, b := range b.AllBlocks() {
-			parent := b.GetMetadata().Parent()
-			if parent != nil {
-				if parentKey != "" && parentKey != parent.String() {
-					parentKey = fmt.Sprintf("%s.%s", parentKey, parent.String())
-				} else {
-					parentKey = parent.String()
-				}
-			}
-			visitor(b, parentKey)
+	if vType == cty.Bool {
+		return val.True(), true
+	}
+
+	return nil, false
+}
+
+type blockVisitor func(block *terraform.Block)
+
+func handleDynamicBlocks(visit blockVisitor) blockVisitor {
```

Comment:
The naming of this is a bit misleading.

Anything like `generateBlockProccessFn` or `createBlockHandlerFn` or something a bit more explicit. I had to go look at the code to understand it has a special condition for `dynamic` but also deals with other blocks.